src/librustc/arena.rs

```diff
@@ -0,0 +1,94 @@
+use arena::{TypedArena, DroplessArena};
+
+#[macro_export]
+macro_rules! arena_types {
+    ($macro:path, $args:tt, $tcx:lifetime) => (
+        $macro!($args, [
+            [] vtable_method: Option<(
+                rustc::hir::def_id::DefId,
+                rustc::ty::subst::SubstsRef<$tcx>
+            )>,
+            [decode] specialization_graph: rustc::traits::specialization_graph::Graph,
+        ], $tcx);
+    )
+}
+
+macro_rules! declare_arena {
+    ([], [$($a:tt $name:ident: $ty:ty,)*], $tcx:lifetime) => {
+        #[derive(Default)]
+        pub struct Arena<$tcx> {
+            dropless: DroplessArena,
+            $($name: TypedArena<$ty>,)*
+        }
+    }
+}
+
+macro_rules! impl_specialized_decodable {
+    ([decode] $ty:ty, $tcx:lifetime) => {
+        impl<$tcx> serialize::UseSpecializedDecodable for &$tcx $ty {}
+    };
+    ([] $ty:ty, $tcx:lifetime) => {};
+}
+
+macro_rules! impl_arena_allocatable {
+    ([], [$($a:tt $name:ident: $ty:ty,)*], $tcx:lifetime) => {
+        $(
+            impl_specialized_decodable!($a $ty, $tcx);
+
+            impl<$tcx> ArenaAllocatable<$tcx> for $ty {
+                #[inline]
+                fn arena<'a>(arena: &'a Arena<$tcx>) -> Option<&'a TypedArena<Self>> {
+                    Some(&arena.$name)
+                }
+            }
+        )*
+    }
+}
+
+arena_types!(declare_arena, [], 'tcx);
+
+arena_types!(impl_arena_allocatable, [], 'tcx);
+
+pub trait ArenaAllocatable<'tcx>: Sized {
+    /// Returns a specific arena to allocate from if the type requires destructors.
+    /// Otherwise it will return `None` to be allocated from the dropless arena.
+    fn arena<'a>(arena: &'a Arena<'tcx>) -> Option<&'a TypedArena<Self>>;
+}
+
+impl<'tcx, T: Copy> ArenaAllocatable<'tcx> for T {
+    #[inline]
+    default fn arena<'a>(_: &'a Arena<'tcx>) -> Option<&'a TypedArena<Self>> {
+        None
+    }
```

Comment:
This impl can conflict with the types in the list :(
```
error[E0119]: conflicting implementations of trait `arena::ArenaAllocatable<'_>` for type `std::collections::HashSet<hir::item_local_id_inner::ItemLocalId, std::hash::BuildHasherDefault<rustc_data_structures::fx::FxHasher>>`:
  --> src\librustc\arena.rs:42:13
   |
4  |  / macro_rules! arena_types {
5  |  |     ($macro:path, $args:tt, $tcx:lifetime) => (
6  |  |         $macro!($args, [
   |  |_________-
7  | ||             [] vtable_method: Option<(
8  | ||                 rustc::hir::def_id::DefId,
9  | ||                 rustc::ty::subst::SubstsRef<$tcx>
...  ||
15 | ||             [] item_local_set: rustc::util::nodemap::ItemLocalSet,
16 | ||         ], $tcx);
   | ||_________________- in this macro invocation
17 |  |     )
18 |  | }
   |  |_- in this expansion of `arena_types!`
...
37 | /  macro_rules! impl_arena_allocatable {
38 | |      ([], [$($a:tt $name:ident: $ty:ty,)*], $tcx:lifetime) => {
39 | |          $(
40 | |              impl_specialized_decodable!($a $ty, $tcx);
41 | |
42 | |              impl<$tcx> ArenaAllocatable<$tcx> for $ty {
   | |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `std::collections::HashSet<hir::item_local_id_inner::ItemLocalId, std::hash::BuildHasherDefault<rustc_data_structures::fx::FxHasher>>`
...  |
49 | |      }
50 | |  }
   | |__- in this expansion of `impl_arena_allocatable!`
...
54 |    arena_types!(impl_arena_allocatable, [], 'tcx);
   |    ----------------------------------------------- in this macro invocation
...
62 |    impl<'tcx, T: Copy> ArenaAllocatable<'tcx> for T {
   |    ------------------------------------------------ first implementation here
   |
   = note: upstream crates may add new impl of trait `std::marker::Copy` for type `std::collections::HashSet<hir::item_local_id_inner::ItemLocalId, std::hash::BuildHasherDefault<rustc_data_structures::fx::FxHasher>>` in future versions

error: aborting due to previous error
```

Do we have some way to express an `ArenaAllocatable` or `Copy` bound?

Reply:
Oh wow, it would be nice to be able to rely on sets and maps never being `Copy`...

Reply:
I found a way around this using a marker trait.